strawberry/experimental/pydantic/utils.py

```diff
@@ -36,3 +40,61 @@ def get_private_fields(cls: Type) -> List[dataclasses.Field]:
             private_fields.append(field)
 
     return private_fields
+
+
+@dataclasses.dataclass()
+class DataclassCreationFields:
+    """Fields required for the fields parameter of make_dataclass"""
+
+    name: str
+    type_annotation: Any
+    field: dataclasses.Field
+
+    @property
+    def to_tuple(self) -> Tuple[str, Any, dataclasses.Field]:
+        # fields parameter wants (name, type, Field)
+        return (self.name, self.type_annotation, self.field)
+
+
+def sort_creation_fields(
+    fields: List[DataclassCreationFields],
+) -> List[DataclassCreationFields]:
+    """
+    Sort fields so that fields with missing defaults go first
+    because dataclasses require that fields with no defaults are defined
+    first
+    """
+    missing_default: List[DataclassCreationFields] = []
+    has_default: List[DataclassCreationFields] = []
+    for model_field in fields:
+        if (
+            model_field.field.default is dataclasses.MISSING
+            and model_field.field.default_factory is dataclasses.MISSING  # type: ignore
+        ):
+            missing_default.append(model_field)
+        else:
+            has_default.append(model_field)
+    return missing_default + has_default
+
+
+def defaults_into_factory(
+    default: Union[_Unset, Any], default_factory: Union[_Unset, NoArgAnyCallable]
+) -> Union[NoArgAnyCallable, _Unset]:
+    """
+    Handle mutable defaults when making the dataclass by using pydantic's smart_deepcopy
+    Returns optionally a NoArgAnyCallable representing a default_factory parameter
+    """
+    final_factory = default_factory
+    if not isinstance(default, _Unset):
+        if not isinstance(default_factory, _Unset):
```

Comment:
Should be able to just compare against `arguments.UNSET`

Reply:
now the problem is that i can't (i think) refer to the singleton instance of arguments.UNSET in the type signature. so instead my type signature has to refer to the _Unset class. And then since my implementation has to be true to my type signature, then i'll need to check for isinstance

Reply:
seems related to this? https://github.com/strawberry-graphql/strawberry/issues/1473

Reply:
```patch
Index: strawberry/experimental/pydantic/utils.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/strawberry/experimental/pydantic/utils.py b/strawberry/experimental/pydantic/utils.py
--- a/strawberry/experimental/pydantic/utils.py	(revision 386b1a367a933996d42713067c7621fa01ff2794)
+++ b/strawberry/experimental/pydantic/utils.py	(date 1640309294688)
@@ -4,7 +4,7 @@
 from pydantic.typing import NoArgAnyCallable
 from pydantic.utils import smart_deepcopy
 
-from strawberry.arguments import _Unset
+from strawberry.arguments import UNSET
 from strawberry.experimental.pydantic.exceptions import (
     BothDefaultAndDefaultFactoryDefinedError,
     UnregisteredTypeException,
@@ -84,8 +84,8 @@
     Returns optionally a NoArgAnyCallable representing a default_factory parameter
     """
     final_factory = default_factory
-    if not isinstance(default, _Unset):
-        if not isinstance(default_factory, _Unset):
+    if default is not UNSET:
+        if default_factory is not UNSET:
             raise BothDefaultAndDefaultFactoryDefinedError(
                 default=default, default_factory=default_factory
             )

```
This change passes tests just fine? Are you seeing something else?

Reply:
that works too. I'm just saying that according to `defaults_into_factory` 's type signature, supposedly you could do this

```
my_new_unset = _Unset()
defaults_into_factory(my_new_unset, my_new_unset)
```

However, the function wouldn't work correctly, because my_new_unset is not strawberry's UNSET. 

Anyways, i don't really care about it. _Unset is a private class by the underscore convention anyways.
so i will make the change. 
Guess you guys could have some singleton restriction in the _Unset to get around the typing restriction.